File: september/week2/sum_of_root_to_leaf_binary_numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, root: TreeNode) -> int:
        stack = [(root, 0)]
        ret = 0
        while stack:
            node, cur_val = stack.pop()
            if node:
                nxt_val = (cur_val << 1) | node.val
                if not node.left and not node.right:
                    ret += nxt_val
                else:
                    stack.append((node.left, nxt_val))
                    stack.append((node.right, nxt_val))
        return ret

File: september/week2/sum_of_root_to_leaf_binary_numbers.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> arr;
    void dfs(TreeNode *root, string cur){
        cur += '0'+root->val;
        if(!(root->left) && !(root->right)){
            arr.push_back(cur);
            return;
        }
        if(root->left)  dfs(root->left,cur);
        if(root->right) dfs(root->right,cur);
    }
    int convert(string tar){
        int ret=0,mul=1;
        for(int i=tar.size()-1;i>=0;i--){
            ret += (tar[i]-'0')*mul;
            mul *= 2;
        }
        return ret;
    }
    int sumRootToLeaf(TreeNode* root) {
        arr.clear();
        arr.resize(0);
        dfs(root,"");
        int ret = 0;
        for(string e: arr)  ret += convert(e);
        return ret;
    }
};
